pkg/benchmark/job/usecase/benchmarkd.go

```diff
@@ -0,0 +1,229 @@
+		service.WithValdClient(c),
+		service.WithJobTypeByString(cfg.Job.JobType),
+		service.WithDimension(cfg.Job.Dimension),
+		service.WithIter(cfg.Job.Iter),
+		service.WithNum(cfg.Job.Num),
+		service.WithMinNum(cfg.Job.MinNum),
+		service.WithRadius(cfg.Job.Radius),
+		service.WithEpsilon(cfg.Job.Epsilon),
+		service.WithTimeout(cfg.Job.Timeout),
+		service.WithHdf5(d),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	h, err := handler.New()
+	if err != nil {
+		return nil, err
+	}
+
+	grpcServerOptions := []server.Option{
+		server.WithGRPCRegistFunc(func(srv *grpc.Server) {
+			// TODO register grpc server handler here
+		}),
+		server.WithGRPCOption(
+			grpc.ChainUnaryInterceptor(recover.RecoverInterceptor()),
+			grpc.ChainStreamInterceptor(recover.RecoverStreamInterceptor()),
+		),
+		server.WithPreStartFunc(func() error {
+			// TODO check unbackupped upstream
+			return nil
+		}),
+		server.WithPreStopFunction(func() error {
+			// TODO backup all index data here
+			return nil
+		}),
+	}
+
+	var obs observability.Observability
+	if cfg.Observability.Enabled {
+		obs, err = observability.NewWithConfig(cfg.Observability)
+		if err != nil {
+			return nil, err
+		}
+		grpcServerOptions = append(
+			grpcServerOptions,
+			server.WithGRPCOption(
+				grpc.StatsHandler(metric.NewServerHandler()),
+			),
+		)
+	}
+
+	srv, err := starter.New(
+		starter.WithConfig(cfg.Server),
+		starter.WithREST(func(sc *iconf.Server) []server.Option {
+			return []server.Option{
+				server.WithHTTPHandler(
+					router.New(
+						router.WithTimeout(sc.HTTP.HandlerTimeout),
+						router.WithErrGroup(eg),
+						router.WithHandler(
+							rest.New(
+							// TODO pass grpc handler to REST option
+							),
+						),
+					)),
+			}
+		}),
+		starter.WithGRPC(func(sc *iconf.Server) []server.Option {
+			return grpcServerOptions
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("pkg/benchmark/job/cmd end")
+
+	return &run{
+		eg:            eg,
+		cfg:           cfg,
+		job:           job,
+		h:             h,
+		server:        srv,
+		observability: obs,
+	}, nil
+}
+
+func (r *run) PreStart(ctx context.Context) error {
+	if r.observability != nil {
+		if err := r.observability.PreStart(ctx); err != nil {
+			return err
+		}
+	}
+	if r.job != nil {
+		return r.job.PreStart(ctx)
+	}
+	return nil
+}
+
+func (r *run) Start(ctx context.Context) (<-chan error, error) {
+	ech := make(chan error, 3)
+	var oech, dech, sech <-chan error
+	r.eg.Go(safety.RecoverFunc(func() (err error) {
+		defer close(ech)
+		if r.observability != nil {
+			oech = r.observability.Start(ctx)
+		}
+		dech, err = r.job.Start(ctx)
+
+		if err != nil {
+			ech <- err
+			return err
+		}
+
+		r.h.Start(ctx)
+
+		sech = r.server.ListenAndServe(ctx)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case err = <-oech:
+			case err = <-dech:
+			case err = <-sech:
+			}
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
```

Comment:
I think an error may lost. so it it better to wrap error or log error? 🤔 

1. wrap error
```go
return errors.Wrap(ctx.Err(), err)
```

2. output error
```go
log.Error(err)
return ctx.Err()

```